### DISCUSSION TITLE
Design Idea : Scenario Mode

### DISCUSSION THREAD

**Discussion by mokun** [mars-sim/mars-sim#426] [Ideas]
This thread is a continuation from [thread 1](https://github.com/mars-sim/mars-sim/issues/213#issuecomment-524014928) about making scenarios.

In [thread 2](https://github.com/mars-sim/mars-sim/issues/86), @shirishag75 talked about various interesting elements of Mars colonization that he has envisioned. 

As this undertaking will take many decades and is multi-generational, how do we make mars-sim more realistic in portraying how it unfolds ? 

As suggested in [thread 3](https://github.com/mars-sim/mars-sim/issues/121), we talked about crafting an *Achievement System*. 

Also, we talked about providing players some composite scores as *Key Performance Indicators* (KPI) that measure the *progress* of how much a settlement has accomplished over time.

Would you agree that probably the next step is to create scenarios that tie all these elements that we have just mentioned together ?

One thing we have discussed is having various **Scenarios**, each of which define some sort of scopes on how settlements would likely interact as well as giving players some concrete goals to work on. 

Feel free to share any scenarios you can think of here.

**Comment by mokun** [comment]
We could start a new xml file called `scenarios.xml`. 

Then we'll need to design and develop what a scenario would look like.

Q: what are some basic attributes or parameters that define the initial states of each scenario ?

Below is a quick example what `scenarios.xml` is like initially.

By all means, let's brainstorm some more.

```
<scenario-list>
    <scenario>
        <scenario-name value="Corporate Expansion" />
        <sponsor value="SpaceX" />
        <base name="Elon Musk City" longitude="20.5 E" latitude="20.9 S" 
        	population="24" bots="48" template="Alpha Base (Phase 4)" />
    </scenario>

    <scenario>
        <scenario-name value="Red Mars" />
        <sponsor value="China" />
        <base name="XingTiang" longitude="4.2 E" latitude="15.9 S" 
                population="4" bots="4" template="Mars Direct Base (Phase 1)" />
    </scenario>
</scenario-list>
```

**Comment by shirishag75** [comment]
For me, it feels more sense to have something like the forward base for e.g. Schillaperi Point (sorry if I got the spelling wrong) which has less number of people for a new commander to handle as an initial scenario. The idea would be to take the player through the learning curve, make her(im) know the concepts and sort of hand-hold them till some sort of tutorial set is over. The hard part would probably be how to do it between the GUI mode and the beryl console for those of us who like the headless version lot more :)

**Comment by mokun** [comment]
> The idea would be to take the player through the learning curve, make her(im) know the concepts and sort of hand-hold them till some sort of tutorial set is over. 

Yes it's a good idea to get players acclimated to the game mechanics by introducing one feature at one time. 

We can discuss how to put together this `Tutorial Mode` on a separate thread.

**Comment by Urwumpe** [comment]
I think we should have a directory "scenarios", with each scenario having its own XML file. ideally one global for the standard scenarios and one in the `.mars-sim` directory of the user. This way, its easier extended by third-party developers.

**Comment by shirishag75** [comment]
> I think we should have a directory "scenarios", with each scenario having its own XML file. ideally one global for the standard scenarios and one in the `.mars-sim` directory of the user. This way, its easier extended by third-party developers.

It might be nice to have something like - 
`
$ java -jar mars-sim-headless-3.1.0-b2-jar-with-dependencies.jar  tutorial1` . The tutorial should have some sort of winnable situation or objective or sub-objectives and the larger objective, so at the end the user 'wins' and can proceed to the next tutorial. 

This way we have have number of tutorials for the user to play and understand the mechanics of the game.  Looking forward to know what people think.

**Comment by Urwumpe** [comment]
Sure, why not? Maybe `--new tutorial1`  to keep the syntax clean? and a `default.xml `that contains the standard starting conditions?

**Comment by shirishag75** [comment]
> Sure, why not? Maybe `--new tutorial1` to keep the syntax clean? and a `default.xml `that contains the standard starting conditions?

+1 just some sort of dialogue, handholding and seeing that there is confirmation of sub-objectives done in the correct manner. The tutorial could also cover skill upgradation as anything the player does should count for skills which make it easier to do it again and make the base more hospitable as well.

**Comment by Urwumpe** [comment]
Well, what about providing a Javascript API to scripts in a scenario file? Java / kotlin both already support this by standard API, other nicer scripting languages like Groovy are also possible by additional dependencies.

**Comment by shirishag75** [comment]
@mokun  please chime in ^

**Comment by mokun** [comment]
> I think we should have a directory "scenarios", with each scenario having its own XML file. ideally one global for the standard scenarios and one in the .mars-sim directory of the user.

Sure we can do that to organize better. 

I know we've been talking about goals. 

So we want to measure the goals in terms of something tangible. What are they ? 

What exactly do we want to specify in a scenario ? 

And what other the initial conditions we may include in the xml file ?

What is the first step we need to change currently in the codebase in order to move foward ? 

btw, we can discuss more about creating dynamic tutorial on this [thread](https://github.com/mars-sim/mars-sim/issues/219) opened by @shirishag75

**Comment by mokun** [comment]
@Urwumpe, we can somewhat tie your [thread](https://github.com/mars-sim/mars-sim/issues/152) to what we would like to achieve with the scenario mode.

How do we justify landing at this coordinate vs. the others ?

Both your thread and this thread come down to the concept of *site selection*.

To make sense of the site the sponsor would like to select, I feel we should talk about how we plan to model where to find any in-situ resources on the surface of Mars, as ISRU will be critical in the long run for any human settlements.

**Comment by Urwumpe** [comment]
Well, I never considered site selection to be more complex than just "I want to go there."  But if we would have some AI decisions there, this would all depend on the factors that are selected for the search and which factors are selected would depend on the ReportingAuthority that picks the base. Or the project that looks for a place for a new base or outpost.

But for scenarios, having fixed coordinates for the initial bases picked by the scenario designer would be good enough right?
